Add Applicative for Arrow

stephen-lazaro opened the pull request:
Adds an `Apply` instance as discussed in https://github.com/typelevel/cats/issues/2069
It can be weakened to `Semigroupal` if we don't buy the `Apply`.
Currently, it is implicit. There was discussion in the issue on whether it should be explicit or implicit. Either answer seems reasonable so I wrote the doctest to be compatible with either way people end up preferring.
Do we need additional tests over the doctest?

kailuowang commented:
This is cool. I start to think `Applicative` is also possible right? 

```scala
def pure[B](a: B): F[A, B] = lift(_ => a) 
```

kailuowang commented on core/src/main/scala/cats/arrow/Arrow.scala:
```diff
@@ -48,3 +48,25 @@ import simulacrum.typeclass
+   * res0: (Long, Int) = (3,6)
+   * }}}
+   */
+  implicit def catsSemigroupalForArrow[F[_, _], A](implicit F: Arrow[F]): Semigroupal[F[A, ?]] = new Apply[F[A, ?]] {
```
it's probably worthwhile to override `product`.

stephen-lazaro replied:
Good catch, will do.

stephen-lazaro commented:
You're totally right, that seems to work. We'll have to check the laws but just from intuition it seems right. I have heard `Arrow is Category + Applicative` so maybe we're now figuring out the `Applicative` part ;)
Will do on the moving and adding the tests. :+1:

julienrf commented:
In case you don’t know this paper, I think it’s a good read about the relationship between arrows and applicatives: http://homepages.inf.ed.ac.uk/slindley/papers/idioms-arrows-monads.pdf

stephen-lazaro commented:
Cool. Strengthened the instance to `Applicative` and added it to the `Applicative` companion object. Also added a test for a function type (let me know if I misunderstood the suggestion there and you'd prefer a stronger test).

kailuowang commented on core/src/main/scala/cats/Applicative.scala:
```diff
@@ -67,8 +68,84 @@ import simulacrum.typeclass
+   * scala> g(5)
+   * res0: String = hello
+   * }}}
+   */
```
Thanks for writing these doctests. But doctests in an instance are probably not necessary. They won't be included in apidoc, and these methods are already thoroughly tested in law tests of the instance. So, although the more tests the better, having them here has two slight cons, a) more code to read, in an instance definition, people want to quickly see how the instance is the implemented, rather than the API usage. and b) they provide redundant test coverage for these implementations, so if somehow the instance is not test by law tests, the test coverage report won't expose that.

stephen-lazaro replied:
Ah, yeah, that makes sense.
